source/pervasive/multiset.rs

```diff
@@ -264,6 +520,57 @@ macro_rules! assert_multisets_equal {
+        forall |m: Multiset<V>, v1: V, mult: nat, v2: V| v1 != v2 ==> #[trigger] m.update(v1, mult).count(v2) == m.count(v2), //lemma_update_different
+        forall |m: Multiset<V>| (#[trigger] m.len() == 0 <==> m =~= Multiset::empty())
+            && (#[trigger] m.len() > 0 ==> exists |v: V| 0 < m.count(v)), //lemma_multiset_empty_len
+        forall |m: Multiset<V>, x: V, y: V| 0 < #[trigger] m.insert(x).count(y) <==> x == y || 0 < m.count(y), //lemma_insert_containment
+        forall |m: Multiset<V>, x: V| #[trigger] m.insert(x).count(x) == m.count(x) + 1, //lemma_insert_increases_count_by_1
+        forall |m: Multiset<V>, x: V, y: V| 0 < m.count(y) ==> 0 < #[trigger] m.insert(x).count(y), //lemma_insert_non_decreasing
+        forall |m: Multiset<V>, x: V, y: V| x != y ==> #[trigger] m.count(y) == #[trigger] m.insert(x).count(y), //lemma_insert_other_elements_unchanged
+        forall |m: Multiset<V>, x: V| #[trigger] m.insert(x).len() == m.len() +1, //lemma_insert_len
+        forall |a: Multiset<V>, b: Multiset<V>, x: V| 
+                #[trigger] a.intersection_with(b).count(x) == min(a.count(x),b.count(x)), //lemma_intersection_count
+        forall |a: Multiset<V>, b: Multiset<V>| #[trigger] a.intersection_with(b).intersection_with(b) == a.intersection_with(b), //lemma_left_pseudo_idempotence
+        forall |a: Multiset<V>, b: Multiset<V>| #[trigger] a.intersection_with(a.intersection_with(b)) == a.intersection_with(b), //lemma_right_pseudo_idempotence
+        forall |a: Multiset<V>, b: Multiset<V>, x: V| #[trigger] a.difference_with(b).count(x) == clip(a.count(x) - b.count(x)), //lemma_difference_count
+        forall |a: Multiset<V>, b: Multiset<V>, x: V| #[trigger] a.count(x) <= #[trigger] b.count(x) 
+                ==> (#[trigger] a.difference_with(b)).count(x) == 0, //lemma_difference_bottoms_out
+{
+    assert forall |m: Multiset<V>, v: V, mult: nat| #[trigger] m.update(v, mult).count(v) == mult by {
+        lemma_update_same(m, v, mult);
+    } 
+    assert forall |m: Multiset<V>, v1: V, mult: nat, v2: V| v1 != v2 implies #[trigger] m.update(v1, mult).count(v2) == m.count(v2) by {
+        lemma_update_different(m, v1, mult, v2);
+    }   
+    assert forall |a: Multiset<V>, b: Multiset<V>, x: V| 
+        #[trigger] a.intersection_with(b).count(x) == min(a.count(x),b.count(x)) by {
+            lemma_intersection_count(a, b, x);
+    } 
+    assert forall |a: Multiset<V>, b: Multiset<V>| #[trigger] a.intersection_with(b).intersection_with(b) == a.intersection_with(b) by {
+        lemma_left_pseudo_idempotence(a, b);
+    }
+    assert forall |a: Multiset<V>, b: Multiset<V>| #[trigger] a.intersection_with(a.intersection_with(b)) == a.intersection_with(b) by {
+        lemma_right_pseudo_idempotence(a, b);
+    }
+    assert forall |a: Multiset<V>, b: Multiset<V>, x: V| #[trigger] a.difference_with(b).count(x) == clip(a.count(x) - b.count(x)) by {
+        lemma_difference_count(a, b, x);
+    }
+}
+
```

Comment:
Sure, my only thought is it's so ugly that it would be best to hide it at the bottom. Like here are all these lemmas and down here is them tied in a little bow. But I'm happy to move it up if you think that fits better with convention.

Reply:
My only thought is that these make more sense being called from the math library, but when I try to call that verus complains that it's not supported. But these functions don't make a whole lot of sense in the Multiset library, which is why I hid them at the bottom. Thoughts on this?

Reply:
Hmm, what error do you see when you put the definitions in the math library?  I think that should be feasible to do.